### DISCUSSION TITLE
Provide building blocks for modal editing

### DISCUSSION THREAD

**Discussion by miquella** [zed-industries/zed#32195] [Feature Requests]
### Check for existing issues

- [X] Completed

### Is your feature request related to a problem?

Very happy that Vim modal editing is being added as a first class feature.

However, editors like Kakoune and Helix have been attempting to make strides in modal editing in recent years. It would be a shame for Zed to be stuck with Vim style editing as the only modal editing solution.

### Describe the solution you'd like

Rather than having Vim input modes and bindings hard-coded in the editor, the editor could allow custom/dynamic "modes" to be created. The Vim bindings could then be built on top of these modes.

If bindings could be built against these custom editing modes, this should open up the possibility of creating new, innovative styles of modal editing that could be used instead of Vim. This would also mean that the bindings could be built outside of the editor, so the other styles and bindings wouldn't need to be included in the editor itself.

**Comment by Daeraxa** [comment]
You might like to see this discussion if you haven't already seen it
https://github.com/zed-industries/feedback/discussions/369

**Comment by wmstack** [comment]
One thing I really want to see is Zed incorporating the selection deselection behaviour of helix. The difficulty is that Vim automatically goes to visual mode as soon as there is any selection, which makes it difficult to create helix like motion.

On the other hand, these shadow selections that are restored when you go back to normal mode, I haven't really found them useful. However the deselection-selection motions in normal mode, I really think it wouldn't be difficult to recreate in Zed.

There has to be a normal-like mode where you can make a selection like a word, line, etc. without going to visual. If I had to give it a name I would call it `select` mode.

Another advantage is that a vim search with `/` will end up picking the first letter of the match, whereas with the suggested `select` mode this can behave like helix by selecting the whole match without leaving `select` mode.

**Comment by notpeter** [comment]
I'm going to move this to a discussion because it's not an actionable feature request / bug to fix.  There has been some ongoing work to support some helix-like functions, but it's relatively slow going.

Practically speaking, we are unlikely to attempt to build an abstraction which allows users to DIY their own modal editor as this issue is suggesting. It's hard enough to faithfully reimplement various motions/actions in all circumstances and requires significant ongoing maintenance and improvement.

Thanks for filing.
